### DISCUSSION TITLE
Example Application: Use of undocumented zephyr/module.yml in application folder

### DISCUSSION THREAD

**Discussion by nylnx** [zephyrproject-rtos/zephyr#49848] [General] [answered]
In the current example application (https://github.com/zephyrproject-rtos/example-application) there is a `zephyr` folder with a `module.yml` file, which enables the use of the device trees, Kconfigs etc. just like in a module.

I cannot find any mention of this mechanism anywhere in the documentation. It might very well be that I haven't fully understood west and its magic, but I really can't find where it says that this is allowed and working inside an application folder of a T2 topology. Or if not specific to the application folder, how it is setup in general.

**Comment by carlescufi** [comment]
Here you go @SkryptX:
https://docs.zephyrproject.org/latest/develop/modules.html#integrate-modules-in-zephyr-build-system

**Comment by nylnx** [comment]
I've read this section multiple times and it does not tell me how the module discovery process works. There is this section https://docs.zephyrproject.org/latest/develop/modules.html#module-inclusion which explains it somewhat, but not completely.

My questions is: How do modules get found? Is it with the `west.yml` and if so, how does the manifest repository of the example-application get added to the list? Is it per directory and if so, why doesn't it recognize my dummy module which satisfies the conditions in the link above and how does it find the modules in the `modules` folder? I'm really at a loss to how it exactly works. I'm very certain I would find _a_ way to do it, but I want to know all the ways and then select the best one depending on my use case.

In essence I'm asking how `ZEPHYR_MODULES` gets constructed and where in the documentation I can find this unambiguously.

**Comment by nylnx** [comment, marked as answer]
I've found my answer why the manifest repository is a "project" and is included in `west list`. See https://github.com/zephyrproject-rtos/west/issues/327

So this feature, that the manifest repository in T2 can simultaneously be a module, is based on the `west list` behavior, which confirmed in the module inclusion link I posted above. 

In my opinion this could be explained more clearly in the documentation that any project listed in the manifest files PLUS the manifest repository are considered projects with all the features they can have, including being a module, as long as they satisfy the conditions mentioned in the module inclusion section.

**Comment by carlescufi** [comment]
@SkryptX you seem to have come to the right conclusions yourself. The module discovery logic, including the manifest repository as a module (eg. example-application) is located here:

https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/zephyr_module.cmake

which in turn invokes this script here:

https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/zephyr_module.py

and that script uses the west APIs (eg. `Manifest.get_projects()` and others) to list all west projects (including the manifest repository) and look for a `zephyr/module.yml` in them.

**Comment by carlescufi** [comment]
@SkryptX we'd be very grateful if you could send a Pull Request with the missing documentation bits, we always welcome contributions! cc @tejlmand.
